src/entry.c

```diff
@@ -1483,3 +1484,119 @@ void
 unlock_entry( const struct stumpless_entry *entry ) {
   config_unlock_mutex( entry->mutex );
 }
+
+/*
+ * static function for stumpless_entry_to_string
+ *  @param dest string to append formatted string
+ *  @param size_t offset dest's offset
+ *  @param key,value it is formatted to key="value",
+ *  @return updated offset
+ */
+static size_t append_key_value_pair(char *dest, size_t offset, const char *key, size_t key_length, const char *value, size_t value_length) {
+  // key="
+  memcpy(dest + offset, key, key_length);
+  offset += key_length;
+  dest[offset++] = '=';
+  dest[offset++] = '"';
+
+  // value
+    memcpy(dest + offset, value, value_length);
+    offset += value_length;
+
+
+  // "
+  dest[offset++] = '"';
+  dest[offset++] = ',';
+
+  return offset;
+}
+
+char *stumpless_entry_to_string(const struct stumpless_entry *entry) {
+    VALIDATE_ARG_NOT_NULL(entry);
+
+    lock_entry(entry);
+
+    const char *app_name = entry->app_name;
+    size_t app_name_length = entry->app_name_length;
+    struct stumpless_element **elements = entry->elements;
+    size_t elements_count = entry->element_count;
+    const char *hostname = entry->hostname;
+    size_t hostname_length = entry->hostname_length;
+    char *message = entry->message;
+    size_t message_length = entry->message_length;
+    const char *message_id = entry->msgid;
+    size_t message_id_length = entry->msgid_length;
+    const char *process_id = entry->procid;
+    size_t process_id_length = entry->procid_length;
+    size_t entry_priority_value = entry->prival;
+
+    unlock_entry(entry);
```

Comment:
This unlock needs to be moved to after the fields of the entry are actually read. The pointers for things like the procid could change if the lock is not held, which would result in a read of memory that is potentially no longer valid.